Use std filesystem instead of Boost.Filesystem if available.

Mike-Devel opened the pull request:
Makes the cmake based boostdep build independent of an installed boost library when you have a recent c++17 toolchain.

Afaik this currently only works with MSVC2017, as gcc-8 needs special linker flags for now and I don't thinkm it is worth supporting that. Unfortunately, I tested it locally, but unfortunately I'm not sure how to hook this up with appveyor CI.

pdimov commented:
What special flags are needed on g++ 8?

Mike-Devel commented:
Afaik you have to explicitly link the filesystem library with '-lstdc++fs'.

pdimov commented:
If that's true, these changes will break `boostdep` on g++ 8.

Mike-Devel commented:
You mean because the cmake file will link to boost, but main.cpp will compile against std::filesystem? 
I think you are right. Sorry, didn't think this through. I'm not at a dev machine right now,  but I can provide a fix tomorrow or the day after.

Mike-Devel commented:
The fix seems straight forward: Don't detect std::filesystem in source, but ifdef on a flag set by the build system. 

I'd consider it worth the trouble to propose the PR, because other than b2, cmake can't trigger the build of boost filesystem on demand (yet). Whether it is worth the trouble for you to  review and maintain it I don't know.

pdimov commented on CMakeLists.txt:
```diff
@@ -4,12 +4,60 @@
+else()
+
+    message(STATUS "Using Boost::filesystem as filesystem library")
+    if( NOT TARGET Boost::filesystem )
```
I'm not fond of `if TARGET` because it's order-dependent.

Mike-Devel replied:
True. Shall I remove it?

Mike-Devel commented:
So, I removed the check for the boost::filesystem target and added a test to travis to make sure my changes don't break gcc 8. Do you think it is worth it to integrate into the main repo? If so, any other changes you want me to make?

pdimov commented:
The fact that `dist/bin/boostdep --list-exceptions` apparently doesn't output anything (https://travis-ci.org/boostorg/boostdep/jobs/449797943) worries me a little, but I don't have time to investigate.

Apart from that, were I to integrate this, I'd probably drop all autodetections and just let the user choose `-DBOOSTDEP_USE_STD_FS`.

Mike-Devel commented:
> The fact that dist/bin/boostdep --list-exceptions apparently doesn't output anything (https://travis-ci.org/boostorg/boostdep/jobs/449797943) worries me a little, but I don't have time to investigate.

Last time I checked, develop and master had the same problem, so I ignored it. It seems to be fixed on develop - will investigate why it still fails on this PR.

pdimov commented:
No, they don't have this problem. The job fails because the exceptions differ, but it doesn't fail - like here - because the output is completely missing.

Mike-Devel commented:
OK, so I found the error: I accidentially inverted an `if <is_folder>` condition.
I also removed the fs autodetection from cmake and simplified the file as much as possible. 

The main remaining problem with this PR - as far as I can tell - is that no CI check actually covers the `std::filesystem` path, as I'm not sure how to best hook this up to appveyor. Do you have a suggestion? Or can you add that check yourself?

Mike-Devel commented:
Ping. 
Is there any interest in this or should I just close it?

pdimov commented:
Keep it open.

Hasn't the CMake community figured out how to link to stdc++fs or c++fs automatically as required?

MikeGitb commented:
Not that I am aware of. Here is some information on the topic: https://gitlab.kitware.com/cmake/cmake/issues/17834

Mike-Devel commented:
Not that I know of. Here is some information on the topic: https://gitlab.kitware.com/cmake/cmake/issues/17834.

Seems it is less a question of how to do it, but more if they should/want to do it.

pdimov commented:
> std::filesystem isn't present in the 7.x branch, and I hope it will be part of libstdc++.so in the 9.x branch, so it should only be needed for 8.x releases.

Sounds like the GCC problem will be going away, but libc++ meanwhile decided to add libc++fs as well.

pdimov commented:
Or maybe not. https://wandbox.org/permlink/tlXI44j9XzL7XUqr

Mike-Devel commented:
Considering that gcc still requires "-latomic" in some scenarios, I won't hold my breath for them to "fix" the `-lstdc++fs` requirement.

Mike-Devel commented:
std::filesystem seems to work with latest trunk versions of clang and gcc (according to the wandbox link you posted). So I guess there is hope.
